accounts/abi/bind/dep_tree.go

```diff
@@ -0,0 +1,140 @@
+	// Contracts/libraries that refer to dependencies in the override
+	// set are linked to the provided address (an already-deployed contract).
+	overrides map[string]common.Address
+}
+
+// NewDeploymentParams instantiates a DeploymentParams instance.
+func NewDeploymentParams(contracts []*MetaData, inputs map[string][]byte, overrides map[string]common.Address) *DeploymentParams {
+	return &DeploymentParams{
+		contracts,
+		inputs,
+		overrides,
+	}
+}
+
+// DeploymentResult encapsulates information about the result of the deployment
+// of a set of contracts: the pending deployment transactions, and the addresses
+// where the contracts will be deployed at.
+type DeploymentResult struct {
+	// map of contract library pattern -> deploy transaction
+	Txs map[string]*types.Transaction
+	// map of contract library pattern -> deployed address
+	Addrs map[string]common.Address
+}
+
+// Accumulate merges `other` into `d`
+func (d *DeploymentResult) Accumulate(other *DeploymentResult) {
+	maps.Copy(d.Txs, other.Txs)
+	maps.Copy(d.Addrs, other.Addrs)
+}
+
+// DeployFn deploys a contract given a deployer and optional input.  It returns
+// the address and a pending transaction, or an error if the deployment failed.
+type DeployFn func(input, deployer []byte) (common.Address, *types.Transaction, error)
+
+// depTreeDeployer is responsible for taking a dependency, deploying-and-linking its components in the proper
+// order.  A depTreeDeployer cannot be used after calling LinkAndDeploy other than to retrieve the deployment result.
+type depTreeDeployer struct {
+	overrideAddrs map[string]common.Address
+	deployedAddrs map[string]common.Address
+	deployerTxs   map[string]*types.Transaction
+	input         map[string][]byte // map of the root contract pattern to the constructor input (if there is any)
+	deploy        DeployFn
+}
+
+// linkAndDeploy recursively deploys a contract and its dependencies:  starting by linking/deploying its dependencies.
+// The deployment result (deploy addresses/txs or an error) is stored in the depTreeDeployer object.
+func (d *depTreeDeployer) linkAndDeploy(metadata *MetaData) error {
+	// don't deploy contracts specified as overrides.  don't deploy their dependencies.
+	if _, ok := d.overrideAddrs[metadata.Pattern]; ok {
+		return nil
+	}
+	// if this contract/library depends on other libraries deploy them (and their dependencies) first
+	for _, dep := range metadata.Deps {
+		if err := d.linkAndDeploy(dep); err != nil {
+			return err
+		}
+	}
+	// if we just deployed any prerequisite contracts, link their deployed addresses into the bytecode to produce
+	// a deployer bytecode for this contract.
+	deployerCode := metadata.Bin
+	for _, dep := range metadata.Deps {
+		linkAddr, ok := d.deployedAddrs[dep.Pattern]
+		if !ok {
+			linkAddr = d.overrideAddrs[dep.Pattern]
+		}
+		deployerCode = strings.ReplaceAll(deployerCode, "__$"+dep.Pattern+"$__", strings.ToLower(linkAddr.String()[2:]))
+	}
```

Comment:
You can combine these two steps, and also make it look nicer if you change the signature to return `common.Address, error`: 
```golang
	// if this contract/library depends on other libraries deploy them (and their dependencies) first
	deployerCode := metadata.Bin
	for _, dep := range metadata.Deps {
		addr, err := d.linkAndDeploy(dep)
		if err != nil {
			return common.Address{}, err
		}
		// link their deployed addresses into the bytecode to produce
		deployerCode = strings.ReplaceAll(deployerCode, "__$"+dep.Pattern+"$__", strings.ToLower(addr.String()[2:]))
	}
```
(still store the addr/tx as previous, though)

Reply:
See second commit in https://github.com/ethereum/go-ethereum/commit/62a780626c913364b947afe4afc82ae1845f65f7

Reply:
Cool. I'm going to pull these in to the branch.

Reply:
BTW, I wouldn't mind if you squash things a bit, perhaps squash sequences of commits from one author into one commit.